File: Perl/2015/day09.pl
#!/usr/bin/env perl 
#===============================================================================
#
#         FILE: day09.pl
#
#        USAGE: ./day09.pl  
#
#  DESCRIPTION: Day 9 - Santa as travelling salesman
#
#      OPTIONS: ---
# REQUIREMENTS: ---
#         BUGS: ---
#        NOTES: ---
#       AUTHOR: Mark T. Johnson (mtj), markj@matzsoft.com
# ORGANIZATION: MATZ Software & Consulting
#      VERSION: 1.0
#      CREATED: 12/08/15 23:50:40
#     REVISION: ---
#===============================================================================

use strict;
use warnings;
use utf8;
use 5.010;

use List::Util qw( min max );


my %map;

while ( <DATA> ) {
    chomp;

    my ( $source, undef, $destination, undef, $distance ) = split;

    $map{$source}{$destination} = $map{$destination}{$source} = $distance;
}

my @routes = routes( keys %map );

say min( @routes );
say max( @routes );



sub routes {
    my @ports = @_;
    my @routes;

    for my $start ( @ports ) {
        push @routes, routes_from( $start, grep { $_ ne $start } @ports );
    }

    return @routes;
}



sub routes_from {
    my $start = shift;
    my @others = @_;
    my @routes;

    if ( @others == 1 ) {
        return ( $map{$start}{ $others[0] } );
    }

    for my $next ( @others ) {
        push @routes, map { $map{$start}{$next} + $_ } routes_from( $next, grep { $_ ne $next } @others );
    }

    return @routes;
}



__END__
Faerun to Tristram = 65
Faerun to Tambi = 129
Faerun to Norrath = 144
Faerun to Snowdin = 71
Faerun to Straylight = 137
Faerun to AlphaCentauri = 3
Faerun to Arbre = 149
Tristram to Tambi = 63
Tristram to Norrath = 4
Tristram to Snowdin = 105
Tristram to Straylight = 125
Tristram to AlphaCentauri = 55
Tristram to Arbre = 14
Tambi to Norrath = 68
Tambi to Snowdin = 52
Tambi to Straylight = 65
Tambi to AlphaCentauri = 22
Tambi to Arbre = 143
Norrath to Snowdin = 8
Norrath to Straylight = 23
Norrath to AlphaCentauri = 136
Norrath to Arbre = 115
Snowdin to Straylight = 101
Snowdin to AlphaCentauri = 84
Snowdin to Arbre = 96
Straylight to AlphaCentauri = 107
Straylight to Arbre = 14
AlphaCentauri to Arbre = 46

File: 2015/day09.swift
//
//         FILE: main.swift
//  DESCRIPTION: day09 - All in a Single Night
//        NOTES: ---
//       AUTHOR: Mark T. Johnson, markj@matzsoft.com
//    COPYRIGHT: © 2021 MATZ Software & Consulting. All rights reserved.
//      VERSION: 1.0
//      CREATED: 07/05/21 21:55:56
//

import Foundation
import Library

struct Map {
    let map: [ String : [ String : Int ] ]
    
    init( lines: [String] ) {
        map = lines.reduce(into: [ String : [ String : Int ] ]() ) { dict, line in
            let words = line.components( separatedBy: " " )
            let distance = Int( words[4] )!
            
            dict[ words[0], default: [:] ][words[2]] = distance
            dict[ words[2], default: [:] ][words[0]] = distance
        }
    }
    
    var routes: [Int] {
        return map.keys.flatMap { name -> [Int] in
            routesFrom( name: name, others: map.keys.filter { $0 != name } )
        }
    }
    
    func routesFrom( name: String, others: [String] ) -> [Int] {
        guard others.count > 1 else { return [ map[name]![others[0]]! ] }
        
        return others.flatMap { next -> [Int] in
            let distance = map[name]![next]!
            return routesFrom( name: next, others: others.filter { $0 != next } ).map { distance + $0 }
        }
    }
}


func parse( input: AOCinput ) -> [Int] {
    return Map( lines: input.lines ).routes
}


func part1( input: AOCinput ) -> String {
    let routes = parse( input: input )
    return "\( routes.min()! )"
}


func part2( input: AOCinput ) -> String {
    let routes = parse( input: input )
    return "\( routes.max()! )"
}


try print( projectInfo() )
try runTests( part1: part1 )
try runTests( part2: part2 )
try solve( part1: part1 )
try solve( part2: part2 )
